modules/internal/ChapelArray.chpl

```diff
@@ -1943,22 +1943,21 @@ module ChapelArray {
   }
 
   @unstable("casting an array to an array type is unstable due to being a new feature — please share any feedback you might have")
-  operator :(arr: [], type t: []) {
-    // Would like to write:
-    //   var a: t = [elem in arr] elem:t.eltType;
-    // but we don't support type queries on array types (yet)
-    // (see https://github.com/chapel-lang/chapel/issues/8543)
-    var a: t;
-
-    if !arr.isRectangular() || !a.isRectangular() {
+  operator :(rhs: [], type t: []) {
+    var res: t;
+
+    if !rhs.isRectangular() || !res.isRectangular() {
       compilerError("Casts between arrays only support rectangular arrays");
-    } else if arr.rank != a.rank {
+    } else if rhs.rank != res.rank {
       compilerError("Casts between arrays require matching ranks");
     } else {
-      a = [elem in arr] elem:a.eltType;
+      type resType = res.eltType;
+      forall (i,j) in zip(res.domain, rhs.domain) {
+        res[i] = rhs[j]:resType;
```

Comment:
What benefit does factoring out `res.eltType` as `resType` get us?

Reply:
Good question.  In one of the previous commits, I was getting it using `res.first.type` and didn't want to compute the first element of `res` in each iteration of the loop, which is why I originally moved it out.

However, while for an array-of-int, this factoring doesn't really add value, in the array-of-arrays-of-... case, `res.eltType` could be a runtime type that requires time to set up and space to store, so factoring it out lets us do that once per loop rather than per iteration.